app/api.go

```diff
@@ -85,89 +92,112 @@ func (lm *LearningMaterialAPI) getOrCreateChallengeEnv() http.HandlerFunc {
 			ErrorResponse(w)
 			return
 		}
-		cc, err := client.GetChallenge(chals)
+		cr, err := client.GetClientRequest(chals)
 
-		//Create a new challenge
+		//Create a new Environment
 		if err != nil {
-			if client.RequestMade() >= lm.conf.API.MaxRequest {
-				ErrorResponse(w) //todo create maxrequest error page
+			if client.RequestMade() >= lm.conf.API.ClientMaxRequest {
+				ClientTooManyRequests(w) //todo create maxrequest error page
 				return
 			}
-			go lm.CreateChallengeEnv(client, mux.Vars(r)["chals"])
+			go lm.CreateEnvironment(client, chals)
+			WaitingResponse(w)
 			return
 		}
 
 		//Check for error while creating the environment
 		select {
-		case err = <-cc.err:
+		case err = <-cr.err:
 			ErrorResponse(w)
 			return
 		default:
 		}
 
-		if !cc.isReady {
-			log.Info().Msgf("[NOT READY] Environment [%s] for the client [%s]", chals, client.ID())
+		if !cr.isReady {
+			//log.Info().Msgf("[NOT READY] Environment [%s] for the client [%s]", chals, client.ID())
 			WaitingResponse(w)
 			return
 		}
 
-		log.Info().Msgf("[READY] Environment [%s] for the client [%s]", chals, client.ID())
+		log.Info().Msgf("[READY] Client Request [%s] for the client [%s]", chals, client.ID())
 
-		authC := http.Cookie{Name: "GUAC_AUTH", Value: cc.guacCookie, Path: "/guacamole/"}
+		authC := http.Cookie{Name: "GUAC_AUTH", Value: cr.guacCookie, Path: "/guacamole/"}
 		http.SetCookie(w, &authC)
-		host := fmt.Sprintf("http://%s:%d/guacamole", lm.conf.Host, cc.guacPort)
+		host := fmt.Sprintf("/guacamole/?%s=%s", requestedChallenges, chals)
+		time.Sleep(5 * time.Second)
 		http.Redirect(w, r, host, http.StatusFound)
 	}
 }
 
-func (lm *LearningMaterialAPI) CreateChallengeEnv(client Client, chals string) {
+func (lm *LearningMaterialAPI) CreateEnvironment(client Client, chals string) {
 
-	envID := GetEnvID(client.ID(), chals)
-	log.Info().Msgf("Creating new Environment [%s]", envID)
+	log.Info().Msgf("Creating new Environment with challenges [%s] for [%s]", chals, client.ID())
 
-	cc := client.NewClientChallenge(chals)
+	cr := client.NewClientRequest(chals)
 
 	chalsTag, _ := lm.GetChallengesFromRequest(chals)
 
-	env, err := lm.newEnvironment(chalsTag, envID)
+	env, err := lm.NewEnvironment(chalsTag)
 	if err != nil {
-		go cc.NewError(err)
+		go cr.NewError(err)
 		return
 	}
```

Comment:
This is a little bit odd to me, the reason is that although error handling has goroutines, core function do not run through goroutines. Ok, it may or may not run with support of goroutines however I can admit that it is a little bit weird usage of goroutines.(-maybe I saw it for the first time :) -)